### PR TITLE
Add test for windup problem in APF

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
The current design for Fair Queueing for Server Requests has a
problem: if the min-max fair result stays different from an even
division for a long time and no queue involved in the imbalance goes
empty then the imbalance keeps accumulating in queue virtual state times.

This commit adds a test that demonstrates the problem.

This commit also generalizes the exerciseQueueSetUniformScenario func so that it can properly evaluate situations where the min-max fair result is not an even division.

It also has some other tweaks to make other tests less flaky.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
PR #91567 attempts to fix this but addresses only one side of the problem.  A deeper fix is needed.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-bugs

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 1]
```diff
@@ -32,6 +34,70 @@ import (
...
+// capacity to the given demands (which slice is not side-effected).
+func FairAlloc(demands []float64, capacity float64) []float64 {
+	lim := len(demands)
+	indices := make([]int, lim)
+	for i := 0; i < lim; i++ {
+		indices[i] = i
+	}
+	sort.Sort(&SortFns{
+		LenFn:  func() int { return lim },
+		LessFn: func(i, j int) bool { return demands[indices[i]] < demands[indices[j]] },
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:57]: Please use https://golang.org/pkg/sort/#Slice instead?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 2]
```diff
@@ -40,107 +106,163 @@ type uniformClient struct {
...
+// in a non-split scenario expectFair1 is what matters.
 // expectedAllRequests indicates whether all requests are expected to get dispatched.
-func exerciseQueueSetUniformScenario(t *testing.T, name string, qs fq.QueueSet, sc uniformScenario,
+func exerciseQueueSetUniformScenario(t *testing.T, name string, qs fq.QueueSet,
+	sc uniformScenario,
+	concurrencyLimit int,
 	evalDuration time.Duration,
-	expectPass, expectedAllRequests, expectInqueueMetrics, expectExecutingMetrics bool,
+	expectFair1, expectFair2, expectedAllRequests bool,
+	evalInqueueMetrics, evalExecutingMetrics bool,
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:137]: This is a lot of parameters and this function is enormous, it might make sense to rephrase this code as an "options" struct with a number of methods. If not in this PR then can we agree to make the next person to touch this, fix it? :) (and add a TODO)

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 3]
```diff
@@ -40,107 +106,163 @@ type uniformClient struct {
...
-// the initial evalDuration and tests to see whether they all got about the same amount.
-// Each client needs to be demanding enough to use this amount, otherwise the fair result
-// is not equal amounts and the simple test in this function would not accurately test fairness.
-// expectPass indicates whether the QueueSet is expected to be fair.
+// the initial evalDuration and tests to see whether they all got about the fair amount.
+// Each client needs to be demanding enough to use more than its fair share,
+// or overall care needs to be taken about timing so that scheduling details
+// do not cause any client to actually request a significantly smaller share
+// than it theoretically should.
+// expectFair1 and expectFair2 indicate whether the QueueSet is expected to be
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:128]: Those are confusing names, my first thought is `testPhaseToFairnessExpectation []bool`, but that's not amazing either. `expectFairPhase1` etc is clearer?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:137]: I'm still reading this function but it is really hard to follow.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:57]: Or, more directly, `sort.Float64s`.  That's where I started.  Then I realized that my problem is not just sorting the input --- the output slice needs to be correlated with the input.  So that's why this code creates an `indices` slice to use for indexing both input and output.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 3 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:128]: I like the slice idea.  How about `expectFair []bool` ?
BTW, I say "expect" rather than "test" or "eval" because there is evaluation in either case, these bools are saying what the expected outcome is.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:57]: Yes, you can't use sort.Float64s, but you can use sort.Slice (on the indicies slice, not the demands slice).

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:57]: Oh, I see.  Yes.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 4]
```diff
@@ -32,6 +34,70 @@ import (
...
+	SwapFn func(i, j int)
+}
+
+func (sc *SortFns) Len() int           { return sc.LenFn() }
+func (sc *SortFns) Less(i, j int) bool { return sc.LessFn(i, j) }
+func (sc *SortFns) Swap(i, j int)      { sc.SwapFn(i, j) }
+
+// FairAlloc computes the min-max fair allocation of the given
+// capacity to the given demands (which slice is not side-effected).
+func FairAlloc(demands []float64, capacity float64) []float64 {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:49]: oh, right

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:137]: I think I can fix both problems together.  Maybe later today.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: privatized

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: factored

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 5]
```diff
@@ -32,158 +34,302 @@ import (
...
+// fair in the respective halves of a split scenario;
+// in a non-split scenario expectFair1 is what matters.
+// expectAllRequests indicates whether all requests are expected to get dispatched.
+type uniformScenario struct {
+	name                                     string
+	qs                                       fq.QueueSet
+	clients                                  []uniformClient
+	concurrencyLimit                         int
+	evalDuration                             time.Duration
+	expectFair                               []bool
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:122]: `[]bool expectFair` is indeed less confusing than `expectFair[1|2]`, while for now  how about we commenting something like: 

`expectFair` is an array of single element indicating whether the concurrency distribution is overall fair for non-split cases. and the on the other hand for split cases, it's a two-element array with `expectFair[0]` indicating whether the splitted half is fair and `expectFair[1]` indicating the overall distribution is fair..

**Review by yue9944882 — COMMENTED** [review]
the test looks good to me, a few minor comments on godoc

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 6]
```diff
@@ -32,158 +34,302 @@ import (
 	"k8s.io/klog/v2"
 )
 
-type uniformScenario []uniformClient
+// fairAlloc computes the min-max fair allocation of the given
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:37]: s/min-max/max-min/?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 7]
```diff
@@ -239,22 +394,33 @@ func TestUniformFlows(t *testing.T) {
...
 	qs := qsc.Complete(fq.DispatchingConfig{ConcurrencyLimit: 4})
 
-	exerciseQueueSetUniformScenario(t, qCfg.Name, qs, []uniformClient{
-		{1001001001, 5, 10, time.Second, time.Second},
-		{2002002002, 5, 10, time.Second, time.Second},
-	}, time.Second*20, true, true, true, true, "", clk, counter)
+	uniformScenario{name: qCfg.Name,
+		qs: qs,
+		clients: []uniformClient{
+			{1001001001, 8, 20, time.Second, time.Second - 1, false},
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:400]: This sort of configuration gets into a cadence where every 1000000000 nanoseconds some requests finish, and 1 ns before that some more arrive.  That means that when one request finishes, the dispatcher already has all the requests it will consider.  If the arrivals were at the same fake time as the completions then we are at the mercy of the golang goroutine scheduler: some dispatching may happen before some of the request arrivals.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 8]
```diff
@@ -32,158 +34,302 @@ import (
...
-	clk *clock.FakeEventClock, counter counter.GoRoutineCounter) {
+// The test measures how much concurrency each client got, on average, over
+// the initial evalDuration and tests to see whether they all got about the fair amount.
+// Each client needs to be demanding enough to use more than its fair share,
+// or overall care needs to be taken about timing so that scheduling details
+// do not cause any client to actually request a significantly smaller share
+// than it theoretically should.
+// expectFair1 and expectFair2 indicate whether the QueueSet is expected to be
+// fair in the respective halves of a split scenario;
+// in a non-split scenario expectFair1 is what matters.
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:114]: good catch, thanks

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:122]: I think it is more meaningful to evaluate the two halves rather than the whole.  The whole will show a lesser degree of whatever effects there may be than an individual half.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:37]: right

**Comment by MikeSpreitzer** [comment]
@rngy

**Comment by MikeSpreitzer** [comment]
The force-push to 7bb15850187 was a rebase.

**Review by yue9944882 — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 9]
```diff
@@ -32,158 +34,302 @@ import (
 	"k8s.io/klog/v2"
 )
 
-type uniformScenario []uniformClient
+// fairAlloc computes the max-min fair allocation of the given
+// capacity to the given demands (which slice is not side-effected).
+func fairAlloc(demands []float64, capacity float64) []float64 {
+	lim := len(demands)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:40]: "count" or "total" for a name?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 10]
```diff
@@ -32,158 +34,302 @@ import (
...
+	alloc := make([]float64, lim)
+	var next int
+	var prevAlloc float64
+	for ; next < lim; next++ {
+		idx := indices[next]
+		demand := demands[idx]
+		if demand <= 0 {
+			continue
+		}
+		hyp := float64(lim-next) * (demand - prevAlloc)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:55]: I can't figure out what `hyp` stands for...

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/integrator.go [hunk 11]
```diff
@@ -101,3 +106,11 @@ func (igr *integrator) GetResults() (results IntegratorResults) {
 	}
 	return
 }
+
+func (igr *integrator) Reset() (results IntegratorResults) {
+	igr.Lock()
+	defer func() { igr.Unlock() }()
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/integrator.go:112]: Looks like this is preexisting, but should just be `defer igr.Unlock()`.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 12]
```diff
@@ -239,22 +394,33 @@ func TestUniformFlows(t *testing.T) {
...
+		expectAllRequests:    true,
+		evalInqueueMetrics:   true,
+		evalExecutingMetrics: true,
+		clk:                  clk,
+		counter:              counter,
+	}.exercise(t)
 }
 
-func TestDifferentFlows(t *testing.T) {
+func TestUniformFlows3(t *testing.T) {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:414]: Can you come up with a more meaningful name than "Flows3"?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 13]
```diff
@@ -32,158 +34,302 @@ import (
...
-					})
-					t.Logf("%s: %d, %d, %d got executed=%v, idle2=%v", clk.Now().Format(nsTimeFmt), i, j, k, executed, idle2)
-					if !executed {
-						atomic.AddUint64(&failedCount, 1)
-						atomic.AddInt32(&rejects[i], 1)
-					}
-				}
-				counter.Add(-1)
-			}(i, j, uc, integrators[i])
+			ust := uniformScenarioThread{
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:166]: This is less deeply nested, so I'm going to assume it's a large improvement :)

**Review by lavalamp — COMMENTED** [review]
This framework is still pretty hard to follow but I think it's better than it was. I made some cosmetic requests.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 10 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:55]: "hypothesis" ... the total usage that would happen if this one gets all it asks for.

**Comment by MikeSpreitzer** [comment]
The force-push to 539c355c8fd is a squash after responding to the latest review comments.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 9 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: `count`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 10 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: I gave this variable a more expansive name and an explanatory comment.  I also commented the loop invariant.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/integrator.go (hunk 11 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/integrator.go:112]: Let me fix this in #91177, which moves this file.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 12 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 10 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go]: Thanks, it's much more understandable now.

**Comment by lavalamp** [comment]
/lgtm
/approve

I think the milestone should be OK as this is an extra unit test.

/milestone v1.19

**Comment by MikeSpreitzer** [comment]
/hold
clearly there is a problem with the pull-kubernetes-node-e2e job now.

**Comment by MikeSpreitzer** [comment]
The force-push to https://github.com/kubernetes/kubernetes/commit/f3fdd5cf9f51f2a03ebfddef82bc2014e04b62a0 is a rebase onto recent `master`.
